File: pkg/data/models/products/details/ram.go
package details

import (
	"BetterPC_2.0/pkg/data/models/products/general"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"strconv"
)

type Ram struct {
	ID           primitive.ObjectID `bson:"_id,omitempty"`
	General      general.General    `bson:"general"`
	Capacity     int                `bson:"capacity"`
	Number       int                `bson:"number"`
	FormFactor   string             `bson:"form_factor"`
	Rank         int                `bson:"rank"`
	Type         string             `bson:"type"`
	Frequency    int                `bson:"frequency"`
	Bandwidth    int                `bson:"bandwidth"`
	CasLatency   string             `bson:"cas_latency"`
	TimingScheme []int              `bson:"timing_scheme"`
	Voltage      float64            `bson:"voltage"`
	Cooling      string             `bson:"cooling"`
	Height       int                `bson:"height"`
}

func (ram Ram) GetProductModel() string {
	return ram.General.Model
}

func (ram Ram) Standardize() general.StandardizedProductData {
	var product general.StandardizedProductData
	product.ProductHeader.ID = ram.ID.Hex()
	product.ProductHeader.ProductType = "ram"
	product.Name = ram.General.Model
	product.General = ram.General
	product.Description = "Capacity: " + strconv.Itoa(ram.Capacity) + "GB, Type: " + ram.Type +
		", Frequency: " + strconv.Itoa(ram.Frequency) + "MHz, CAS Latency: " + ram.CasLatency
	return product
}
func (ram Ram) ProductFinalPrice() int {
	return ram.General.CalculateFinalPrice()
}

type UpdateRamInput struct {
	General      *general.General `bson:"general"`
	Capacity     *int             `bson:"capacity"`
	Number       *int             `bson:"number"`
	FormFactor   *string          `bson:"form_factor"`
	Rank         *int             `bson:"rank"`
	Type         *string          `bson:"type"`
	Frequency    *int             `bson:"frequency"`
	Bandwidth    *int             `bson:"bandwidth"`
	CasLatency   *string          `bson:"cas_latency"`
	TimingScheme *[]int           `bson:"timing_scheme"`
	Voltage      *float64         `bson:"voltage"`
	Cooling      *string          `bson:"cooling"`
	Height       *int             `bson:"height"`
}

func (r UpdateRamInput) Validate() error {
	return general.ValidateStruct(&r)
}

File: frontend/better_pc/src/modules/products/models/ram.ts
import {BaseProduct} from "@/modules/products/models/product_common"; // Общий интерфейс General

export interface Ram extends BaseProduct {
    capacity: number; // Объем памяти (в гигабайтах)
    number: number; // Количество модулей
    formFactor: string; // Форм-фактор RAM (например, DIMM, SO-DIMM)
    rank: number; // Ранг памяти (например, одиночный или двухканальный)
    type: string; // Тип памяти (например, DDR4, DDR5)
    frequency: number; // Частота работы памяти (например, 3200 MHz)
    bandwidth: number; // Пропускная способность памяти (например, 25.6 GB/s)
    casLatency: string; // CAS латентность (например, 16-18-18)
    timingScheme: number[]; // Тайминги памяти в виде массива (например, [16, 18, 18, 36])
    voltage: number; // Напряжение работы памяти (например, 1.35V)
    cooling: string; // Охлаждение памяти (например, с радиатором, без охлаждения)
    height: number; // Высота модуля памяти (например, в миллиметрах)
}
